Improve mason update output

Aniket21mathur opened the pull request:
Modifies the "Updating mason-registry" message to specify the registry name, for example:

Using the chapel-lang registry and a custom registry changes from this:

```
Updating mason-registry
Updating mason-registry
```

to this:

```
Updating registry
Updating myRegistry
```
Fixes https://github.com/chapel-lang/chapel/issues/13336

Aniket21mathur commented:
@ben-albrecht I am trying to run a mason test file `mason-registry.chpl` with command `chpl mason-registry.chpl`. But I am getting the following error:

```
mason-registry.chpl:3: error: Cannot find module or enum 'MasonEnv'
```
I had already built mason with `make mason` in `$CHPL_HOME`.

I also tried with command `mason run mason-registry.chpl`. But I am getting the following error:

`Mason could not find your configuration file (Mason.toml)
`
I went through the [documentation](https://chapel-lang.org/docs/tools/mason/mason.html#installation-instructions), as per it a `Mason.toml` is included in a new package. So how should we compile the tests?

Need help.

Thanks!

ben-albrecht commented:
@Aniket21mathur - the test system (`start_test`) utilizes all the metadata files laying around in that directory. One of those is `COMPOPTS` which contains the necessary compiler flags to compile the example. You'll have to throw those manually if you want to build the program outside of the test framework. See the test framework docs for more details.

ben-albrecht commented on tools/mason/MasonUpdate.chpl:
```diff
@@ -165,7 +165,7 @@ proc updateRegistry(tf: string, args: list(string)) {
       mkdir(localRegistry, parents=true);
       const cloneRegistry = 'git clone -q ' + registry + ' .';
-      writeln("Updating mason-registry");
+      writeln("Updating local-registry");
```
Each registry directory has its own name. This change will require:

1. Getting that registry name in a variable
2. Printing that registry name here, instead of `mason-registry`

Aniket21mathur replied:
> 2\. Printing that registry name here, instead of `mason-registry`

Ohh! Seems like I misunderstood the issue. I will update the changes accordingly :+1:

Aniket21mathur replied:
@ben-albrecht I think we already have the registry name as `registry`, we just need to print it?

ben-albrecht replied:
oh, possibly. Maybe this is still an easy fix :)

Aniket21mathur replied:
I have made the changes. Please have a look. :smile:

Aniket21mathur commented:
> Didn't seem to work as expected:

Yes, do not work for me as well. Looks strange. The part of the `updateRegistry` proc that gives these outputs is as follows

```
if isDir(registryHome) {
      var pullRegistry = 'git pull -q origin master';
      if tf == "Mason.toml" then
        writeln("Updating mason-registry");
      gitC(registryHome, pullRegistry);
    }
    // Registry has moved or does not exist
    else {
      mkdir(MASON_HOME + '/src', parents=true);
      const localRegistry = registryHome;
      mkdir(localRegistry, parents=true);
      const cloneRegistry = 'git clone -q ' + registry + ' .';
      writeln("Updating ", registry, "-registry");
      gitC(localRegistry, cloneRegistry);
    }
```
And I have made changes to this part only. For testing that the changes made to the code are reflected or not I did something like

```
        writeln("Updating mason-registry");
        writeln("foo");
```
But `foo` is not printed while testing the way you did above :confused:

ben-albrecht commented:
There must be another place where we print the `Updating mason-registry` then.

Aniket21mathur commented:
@ben-albrecht I did a global search in `atom` text editor with keyword `Updating mason-registry` (in this branch) and I got the following output

![Screenshot from 2020-02-13 01-10-05](https://user-images.githubusercontent.com/35092582/74370765-a647a800-4dfd-11ea-9b41-0acce1ee6335.png)

I found the keyword only in 2 directories, one is `test` and the other is `tools`(`tools/mason/MasonUpdate.chpl`).

I think any changes to the test directory will not work (please correct me if I am wrong) so we are left with only one directory.

Please guide. Thanks!

ben-albrecht commented:
> But foo is also not printed while testing the way you did above 😕

I just checked and confirmed those lines in `MasonUpdate.chpl` are correct. Are you rebuilding mason after modifying them, i.e. `make mason`?

Aniket21mathur commented:
> Are you rebuilding mason after modifying them, i.e. `make mason`?

@ben-albrecht thanks for pointing out. After making the changes and building mason again. I got the expected output

```
Updating mason-registry
Updating /home/user/chapel/bens-registry-registry
```


> Didn't seem to work as expected:

I am not sure why it happened. Did you rebuild mason?

Aniket21mathur commented:
> Are you able to update this such that we don't add `-registry` the registry names?

@ben-albrecht I have made the desired changes. Please review :smile:

Aniket21mathur commented:
@ben-albrecht thanks for testing.

Currently, we are getting the following output:

```
Updating mason-registry
Updating /path/to/registries/bens-registry
```
And we aim to get

```
Updating mason-registry
Updating bens-registry
```


> 1. The `updating <registry>` message should use the identifier defined in `MASON_REGISTRY`. For example:

Sorry, I am not familiar with identifiers in `chapel` and thus this statement is not that clear to me.
How exactly we use the identifier with the message?

> 1. We need to modify the "Updating mason-registry" print on line 159 as well

This is also not clear to me, what exactly we modify this message to?

Please help. Thanks!

ben-albrecht commented:
Sorry to be unclear.

> Sorry, I am not familiar with identifiers in chapel and thus this statement is not that clear to me.
How exactly we use the identifier with the message?

What I mean is that `MASON_REGISTRY` should contain a list of tuples, with values: `[(mason-registry, <URL>), (bens-registry, <URL>)]`, so you can get the values from there rather than re-implementing code to parse the names out of the URLs/paths.


>> We need to modify the "Updating mason-registry" print on line 159 as well
> This is also not clear to me, what exactly we modify this message to?

Line 159 also prints a hard-coded `"mason-registry"`. We should update that message to use the variable as well.

Aniket21mathur commented:
@ben-albrecht I have made the desired changes.

I also found something strange. The following code works fine

```


const env = "mason-registry|https://github.com/chapel-lang/mason-registry,bens-registry|/home/h4ck3y3/chapel/bens-registry";
const a = env.split(',');
const b = a[2].split('|');
writeln(b[1]);
```
Output

```
bens-registry
```
which was expected

But If the code is modified as below

```
const env = "mason-registry|https://github.com/chapel-lang/mason-registry,bens-registry|/home/h4ck3y3/chapel/bens-registry";
const a = env.split(',')[2].split('|');
writeln(a[1]);
```
Output

```
a
```
Expected

```
bens-registry
```

Please have a look. Thanks!

ben-albrecht commented:
Sorry, I still was unclear. I meant the `MASON_REGISTRY` Chapel variable, which already handles the splitting of the environment variable string. I will comment inline with a specific suggestion.

ben-albrecht commented on tools/mason/MasonUpdate.chpl:
```diff
@@ -153,10 +153,17 @@ proc updateRegistry(tf: string, args: list(string)) {
   checkRegistryChanged();
   for ((name, registry), registryHome) in zip(MASON_REGISTRY, MASON_CACHED_REGISTRY) {
```
We are actually already looping over `MASON_REGISTRY` tuple values, so the `name` is what we want to print. on the "Updating" lines.

Go check out the definition of `MASON_REGISTRY` (a parens-less procedure, so search `proc MASON_REGISTRY`), for reference.

Aniket21mathur commented:
> Sorry, I still was unclear. I meant the `MASON_REGISTRY` Chapel variable, which already handles the splitting of the environment variable string. I will comment inline with a specific suggestion.

Thanks for explaining, it was my mistake. I got a bit confused :sweat_smile: 

I made the desired changes and tested them.

Aniket21mathur commented:
> Looks good! Can you verify that testing passes (`test/mason`)?

I have made some changes to the `good` files, that are required as now we have updated the message.

All local tests pass now :-)

```
[Test Summary - 200215.011925]
[Summary: #Successes = 75 | #Failures = 0 | #Futures = 0 | #Warnings = 0 ]
[Summary: #Passing Suppressions = 0 | #Passing Futures = 0 ]
[END]
```
Thanks!

Aniket21mathur commented:
@ben-albrecht just a reminder, no hurry 😁

ben-albrecht commented:
@Aniket21mathur - note (for future PRs) that I updated the PR description to give more information. Thanks again!

Aniket21mathur commented:
> @Aniket21mathur - note (for future PRs) that I updated the PR description to give more information.

Thanks. Noted :smile:
